R/Chrome.R

```diff
@@ -2,6 +2,141 @@
 #' @importFrom assertthat assert_that is.scalar is.number
 NULL
 
+#' Execute an asynchronous CDP flow with Chrome
+#'
+#' The `chrome_execute()` function executes an asynchronous Chrome DevTools
+#' Protocol flow with Chromium/Chrome and can turn it into a synchronous function.
+#' An asynchronous remote flow is a function that takes a connection object and
+#' returns a [promise][promises::promise].
+#' If several functions are passed to `chrome_executes()`, their execution is
+#' serial. If one of the asynchronous functions fails, the whole execution also
+#' fails.
+#'
+#' @param ... Asynchronous remote flow functions.
+#' @param .list A list of asynchronous remote flow functions - an alternative to
+#' `...`.
+#' @param timeouts A vector of timeouts applied to each asynchronous function.
+#'     Repeated.
+#' @param cleaning_timeout The delay for cleaning Chrome.
+#' @param async Is the result a promise? Required for using `chrome_execute()`
+#'     in Shiny.
+#' @param bin Character scalar, the path to Chromium or Chrome executable.
+#' @param debug_port Integer scalar, the Chromium/Chrome remote debugging port.
+#' @param local Logical scalar, indicating whether the local version of the
+#'     protocol (embedded in `crrri`) must be used or the protocol must be
+#'     fetched _remotely_.
+#' @param extra_args Character vector, extra command line arguments passed to
+#'     Chromium/Chrome.
+#' @param headless Logical scalar, indicating whether Chromium/Chrome is launched
+#'     in headless mode.
+#' @param retry_delay Number, delay in seconds between two successive tries to
+#'     connect to headless Chromium/Chrome.
+#' @param max_attempts Logical scalar, number of tries to connect to headless
+#'     Chromium/Chrome.
+#'
+#' @return An invisible list with the values of the fulfilled promises for each
+#'     async function.d If there is only async function, the return value is the value of the
+#'     fulfilled promise.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' async_save_as_pdf <- function(url) {
+#'   function(client) {
+#'     Page <- client$Page
+#'
+#'     Page$enable() %...>% {
+#'       Page$navigate(url = url)
+#'       Page$loadEventFired()
+#'     } %...>% {
+#'       Page$printToPDF()
+#'     } %...>% {
+#'       .$data %>%
+#'         jsonlite::base64_dec() %>%
+#'         writeBin(paste0(httr::parse_url(url)$hostname, ".pdf"))
+#'     }
+#'   }
+#' }
+#'
+#' save_as_pdf <- function(...) {
+#'   list(...) %>%
+#'     purrr::map(async_save_as_pdf) %>%
+#'     chrome_execute(.list = .)
+#' }
+#'
+#' save_as_pdf("https://www.r-project.org/", "https://rstudio.com/")
+#' }
+chrome_execute <- function(
+  ..., .list = NULL, timeouts = 30, cleaning_timeout = 30, async = FALSE,
+  bin = Sys.getenv("HEADLESS_CHROME"), debug_port = 9222L, local = FALSE,
+  extra_args = NULL, headless = TRUE, retry_delay = 0.2, max_attempts = 15L
+) {
+  if(is.null(.list)) {
+    funs <- list(...)
+  } else {
+    stopifnot(is.list(.list))
```

Comment:
should we keep on going with `assertthat` ? 
```suggestion
    assert_that(rlang::is_list(.list))
```

Reply:
As you prefer, I find that the message returned by `assert_that(rlang::is_list(.list))` is uglier (and not translated). Maybe I was too lazy to implement the whole `assertthat` stuff.
I'll do that.

Reply:
You put custom message for other. I think we should keep on with the same stuff instead of switching between stopifnot and assertthat 
I can do it, no problem

Reply:
> (and not translated)

I see what you mean now by translated. I raises the question of when we want to use `stopifnot` and when to use `assert_that`. It could be weird to have some error translated and other not translated. 

For going with other message, I would have gone with
``` r
is_list <- function(x) {
  rlang::is_list(x)
}

assertthat::on_failure(is_list) <- function(call, env) {
  paste0(deparse(call$x), "must be a list")
}
assertthat::assert_that(is_list("a"))
#> Error: "a"must be a list
```

But at the end I think 

* `assertthat` error should be reserved for input checking
* `stopifnot` or other, should be use internally inside functions when we need to throw an error.  Know that we have `rlang::abord` that would allow us to create custom handler for specific error. We have lots of things going on with several object, remote connection, chrome events, etc..
I think it could be useful at some point. 

What do you thing of all that ?

Reply:
Sound good to me! :+1:

Reply:
done in https://github.com/RLesur/crrri/pull/62/commits/b1b754ac28ed6feb666fab577af636ac828840b2